eyelid-match-ops/src/primitives/poly.rs

```diff
@@ -184,24 +196,28 @@ fn rec_karatsuba_mul_inner(a: &Poly, b: &Poly, chunk: usize) -> Poly {
 // - split the `for` and `while` loops into functions, and benchmark the overall performance.
 // - split large code blocks into smaller functions, and benchmark the overall performance.
 #[allow(clippy::cognitive_complexity)]
-pub fn flat_karatsuba_mul(a: &Poly, b: &Poly) -> Poly {
+pub fn flat_karatsuba_mul<const MAX_POLY_DEGREE: usize>(
+    a: &Poly<MAX_POLY_DEGREE>,
+    b: &Poly<MAX_POLY_DEGREE>,
+) -> Poly<MAX_POLY_DEGREE> {
     debug_assert!(a.degree() <= MAX_POLY_DEGREE);
     debug_assert!(b.degree() <= MAX_POLY_DEGREE);
 
-    /// The final number of layers in the flat Karatsuba `while` loop.
-    /// `FLAT_KARATSUBA_INITIAL_LAYER` skips some layers.
-    const RECURSION_HEIGHT: u32 = usize::ilog2(MAX_POLY_DEGREE);
-    const_assert!(FLAT_KARATSUBA_INITIAL_LAYER <= RECURSION_HEIGHT);
+    // The final number of layers in the flat Karatsuba `while` loop.
+    // `FLAT_KARATSUBA_INITIAL_LAYER` skips some layers.
+    let recursion_height: u32 = usize::ilog2(MAX_POLY_DEGREE);
+
+    debug_assert!(FLAT_KARATSUBA_INITIAL_LAYER <= recursion_height);
```

Comment:
I think a change like this might be out of scope for this PR, which is focused on adding a generic modulus degree to the polynomial type.

This specific change is needed because Rust limits usage of const generics in const code. (They are a different kind/scope of const.) So I just used the same kind of asserts as existing code.

We could work through this change and check its performance in another ticket?

Reply:
> * to help the compiler make better code. E.g., Vec range checks can be removed from the hot code because sizes were checked before the loop.

There's a tradeoff here between:
* small production code size, limited branching, and the optimisations they enable, and
* the trivial CPU cost of assertions, and the optimisations they enable.

Unfortunately, the only way to be sure of the best tradeoff is to benchmark both implementations. I've seen it go both ways.

Rust can also hoist range checks out of the loop independently of assertions. But some of these optimisations only apply to small loops. (Or small functions.)

Reply:
Moved to issue https://github.com/Inversed-Tech/eyelid/issues/58